src/main/java/com/amazon/ion/impl/IonCursorBinary.java

```diff
@@ -1568,6 +1568,9 @@ private boolean uncheckedNextToken() {
             if (uncheckedNextContainedToken()) {
                 return false;
             }
+            if (peekIndex >= limit) {
+                throw new IonException("Malformed data: declared length exceeds the number of bytes remaining in the stream.");
```

Comment:
Would
> ...exceeds the number of bytes remaining **in the container**.

be more accurate?

Reply:
It took me a while to understand why you made this suggestion. It's because we're in an unchecked method, which means that we already believe we have enough data to finish the container. Therefore if `peekIndex >= limit` we're definitely past the container bounds because implicitly `parent.endIndex < limit`?

Another thing strikes me about this though- we check the same condition `peekIndex >= limit` on line 1551, and from my read the only thing that can change that condition in between there and here is this clause of `uncheckedNextContainedToken()`:

```java
        } else if (parent.typeId.type == IonType.STRUCT) {
            if (minorVersion == 0) {
                byte b = buffer[(int) peekIndex++];
                if (b < 0) {
                    fieldSid = (b & LOWER_SEVEN_BITS_BITMASK);
                } else {
                    fieldSid = (int) uncheckedReadVarUInt_1_0(b);
                }
            } else {
                uncheckedReadFieldName_1_1();
            }
        }
```

We must be in `minorVersion == 0`, and I note that `uncheckedReadVarUInt_1_0(b)` already contains a malformed data check. Does that mean that this error condition is discovered only when we've just read a 1-byte VarUInt field name, in the ` if (b < 0) {` case above? In that case, should this error check go there, in `uncheckedNextContainedToken()`?

Reply:
> Would
> 
> > ...exceeds the number of bytes remaining **in the container**.
> 
> be more accurate?

Yes, this is more accurate.

> We must be in minorVersion == 0, and I note that uncheckedReadVarUInt_1_0(b) already contains a malformed data check. Does that mean that this error condition is discovered only when we've just read a 1-byte VarUInt field name, in the  if (b < 0) { case above? In that case, should this error check go there, in uncheckedNextContainedToken()?

I preferred to put this check in the proposed location, rather than in `uncheckedNextContainedToken()`, because the check protects the line that immediately follows (the access to the buffer at `peekIndex`). Note: `uncheckedReadVarUInt_1_0` performs the check *before* each byte it consumes, but the added check protects access to the first byte *after* the field name.